Add binding and kind to drgn.Symbol

brenns10 opened the pull request:
This branch implements #109 -- better late than never!

osandov commented on libdrgn/drgn.h.in:
```diff
@@ -2566,6 +2566,49 @@ struct drgn_error *drgn_format_type(struct drgn_qualified_type qualified_type,
+	DRGN_SYMBOL_KIND_COMMON,
+	/** Symbol is thread-local data object */
+	DRGN_SYMBOL_KIND_TLS,
+	/* The above are ordered to match elf.h, do not change. */
```
Likewise, this also applies to `DRGN_SYMBOL_KIND_IFUNC`?

osandov commented on libdrgn/python/drgnpy.h:
```diff
@@ -162,6 +162,8 @@ extern PyObject *PrimitiveType_class;
 extern PyObject *ProgramFlags_class;
 extern PyObject *Qualifiers_class;
 extern PyObject *TypeKind_class;
+extern PyObject *SymbolBinding_class;
+extern PyObject *SymbolKind_class;
```
Also alphabetical here.

osandov commented on libdrgn/symbol.c:
```diff
@@ -13,6 +14,36 @@ LIBDRGN_PUBLIC void drgn_symbol_destroy(struct drgn_symbol *sym)
+	ret->address = address;
+	ret->size = elf_sym.st_size;
+	binding = GELF_ST_BIND(elf_sym.st_info);
+	if (binding < 0 || binding >= STB_NUM) {
+		if (binding == STB_GNU_UNIQUE)
+			ret->binding = DRGN_SYMBOL_BINDING_UNIQUE;
+		else
+			ret->binding = DRGN_SYMBOL_BINDING_UNKNOWN;
+	} else {
+		/* offset by 1 from elf.h definitions */
+		ret->binding = binding + 1;
+	}
```
We should compare against the largest constant that we actually know about, just in case `elf.h` gets new constants and raises `STB_NUM`.

`ST_BIND` is defined as `(unsigned char)st_info >> 4`, so binding will always be non-negative. We can simplify it a bit further:

```suggestion
	if (binding <= STB_WEAK || binding == STB_GNU_UNIQUE)
		ret->binding = binding + 1;
	else
		ret->binding = DRGN_SYMBOL_BINDING_UNKNOWN;
```

osandov commented on libdrgn/symbol.c:
```diff
@@ -13,6 +14,36 @@ LIBDRGN_PUBLIC void drgn_symbol_destroy(struct drgn_symbol *sym)
+		ret->binding = binding + 1;
+	}
+	kind = GELF_ST_TYPE(elf_sym.st_info);
+	if (kind < 0 || kind >= STT_NUM) {
+		if (kind == STT_GNU_IFUNC)
+			ret->kind = DRGN_SYMBOL_KIND_IFUNC;
+		else
+			ret->kind = DRGN_SYMBOL_KIND_NOTYPE;
+	} else {
+		/* no offset from elf.h definitions */
+		ret->kind = kind;
+	}
```
Same idea here:

```suggestion
	if (kind <= STT_TLS || kind == STT_GNU_IFUNC)
		ret->kind = kind;
	else
		ret->kind = DRGN_SYMBOL_KIND_UNKNOWN;
```

osandov commented on libdrgn/symbol.h:
```diff
@@ -5,11 +5,20 @@
 
+/** Populates the fields of @ref drgn_symbol. */
+void drgn_symbol_init_from_elf(const char *name, uint64_t address,
+			       GElf_Sym elf_sym, struct drgn_symbol *ret);
```
`GElf_Sym` is 24 bytes, so I think we should pass it as a pointer rather than by value.

```suggestion
			       const GElf_Sym *elf_sym, struct drgn_symbol *ret);
```

osandov commented on libdrgn/symbol.c:
```diff
@@ -13,6 +14,36 @@ LIBDRGN_PUBLIC void drgn_symbol_destroy(struct drgn_symbol *sym)
 	free(sym);
 }
 
+LIBDRGN_PUBLIC void
```
This is internal to libdrgn, so it shouldn't be exported with `LIBDRGN_PUBLIC`.

osandov commented on libdrgn/symbol.h:
```diff
@@ -5,11 +5,20 @@
 #define DRGN_SYMBOL_H
 
 #include <stdint.h>
```
You can delete this since `drgn.h` is documented as including it for you.

```suggestion
```

osandov reviewed (changes requested):
Thanks, this looks great! I have mostly nitpicks. It'd also be nice to have test cases for this, but I'll have to think about how to do that. Maybe lump it in with the Linux kernel helper tests and test it against some symbols in `/proc/kallsyms`.

brenns10 commented on _drgn.pyi:
```diff
@@ -1953,6 +1959,51 @@ class TypeKind(enum.Enum):
+    """The symbol is unique among all global and local symbols."""
+
+class SymbolKind(enum.Enum):
+    "A ``SymbolKind`` represents what sort of value a symbol represents."
```
I waffled over the word choice. You're right, value is typically used differently in the context of programming/debugging. I didn't want to use Object due to the overloading, since really only one of these (DATA) would be an object in the technical sense - others would be functions or something else entirely. Maybe "entity" works? `A SymbolKind describes what sort of entity a symbol represents`
(also getting rid of the double "represents"...)

brenns10 replied:
Resolving this, feel free to reopen if you disagree with entity.

brenns10 commented on libdrgn/drgn.h.in:
```diff
@@ -2566,6 +2566,49 @@ struct drgn_error *drgn_format_type(struct drgn_qualified_type qualified_type,
+	DRGN_SYMBOL_BINDING_GLOBAL,
+	/** A weak symbol, which could be overridden. */
+	DRGN_SYMBOL_BINDING_WEAK,
+	/* The above are ordered to match elf.h (offset by 1), do not change. */
```
I meant the comment to warn against reordering these, since they're implicitly numbered. `DRGN_SYMBOL_BINDING_UNIQUE` is separate and explicitly numbered so doesn't have a risk. How about I change the comment to read `do not reorder`? (same below)

brenns10 replied:
Resolving this, feel free to reopen if you don't like the rewording.

brenns10 commented on _drgn.pyi:
```diff
@@ -1953,6 +1959,51 @@ class TypeKind(enum.Enum):
+class SymbolKind(enum.Enum):
+    "A ``SymbolKind`` represents what sort of value a symbol represents."
+
+    NOTYPE = ...
```
I think that makes sense, since I did overload the NOTYPE value with the meaning of UNKNOWN. I'll admit I don't have any concept of what NOTYPE gets used for, so I'd also be willing to create a separate UNKNOWN so we can differentiate. But I'll stick with renaming it for now.

brenns10 commented:
I went ahead and addressed all your comments, here's the diff between versions since I force-pushed it to keep git history clean.

https://github.com/osandov/drgn/compare/5d86b01594545fed1a5e4e3dfd95b24cd6be8a34..52040c69ea71bc092c4e5b09bc2414bce7e5d567

I agree on testing, something automatic would be nice. I think I need to read through the test code already here and I can come up with a suggestion.

Here's my quick manual test on a vmcore I had laying around.
```
>>> stext = prog.symbol("_stext")
>>> stext
Symbol(name='_stext', address=0xffffffff92600000, size=0x0)](url)
>>> stext.kind
<SymbolKind.UNKNOWN: 0>
>>> stext.binding
<SymbolBinding.GLOBAL: 2>
>>> 
>>> schedule = prog.symbol("schedule")
>>> schedule
Symbol(name='schedule', address=0xffffffff92fd92e0, size=0x9a)
>>> schedule.kind
<SymbolKind.FUNC: 2>
>>> schedule.binding
<SymbolBinding.GLOBAL: 2>
```

This raises a question -- should I adjust the `Symbol()` repr to include the kind and binding?

osandov commented:
We can probably count on `jiffies` being a global object, so something like this would probably work as a minimal test case (in a new `tests/helpers/linux/test_symbol.py` file):

```python
class TestSymbol(LinuxHelperTestCase):
    def test_global_symbol(self):
        symbol = self.prog.symbol("jiffies")
        self.assertEqual(symbol.name, "jiffies")
        self.assertEqual(symbol.binding, SymbolBinding.GLOBAL)
        self.assertEqual(symbol.kind, SymbolKind.OBJECT)
```

And yes that's a good catch, we should add the binding and kind to repr.

brenns10 commented:
Added the binding and kind to Symbol's repr, using the default enum reprs. It doesn't look perfect (i.e. you couldn't paste it into a python session to get a new instance of the same object) but I think that's expected with nested data structures?

```
>>> prog.symbol("jiffies")
Symbol(name='jiffies', address=0xffffffff93a06980, size=0x0, binding=<SymbolBinding.GLOBAL: 2>, kind=<SymbolKind.OBJECT: 1>)
```

Anyway, I also ran the test and it passes (CI should verify too).

osandov reviewed (approved):
I made a few cosmetic changes:

* Moved `SymbolBinding` and `SymbolKind` in `_drgn.pyi` next to `Symbol`.
* Clarified some of the bindings and kinds, and added references for the obscure GNU extensions.
* Added `SymbolBinding` and `SymbolKind` to `docs/api_reference.rst`.
* Reworded some of the comments in `libdrgn/drgn.h.in`.
* Moved `Symbol_repr` to avoid the need for a forward declaration.
* A couple of small coding style nits.

This should be ready to go once tests pass. Thanks again!
